### DISCUSSION TITLE
Why is this schema validation working?

### DISCUSSION THREAD

**Discussion by karl0ss** [python-jsonschema/jsonschema#1293] [Q&A]
I have taken an openapi json file and am loading that to do schema validation against some api's...here is my sample test code -

```
import json
import requests
from jsonschema import validate, ValidationError
from jsonref import JsonRef


with open("openapi.json", "r") as file:
    openapi_spec = JsonRef.replace_refs(json.load(file))

paths = [
    "/Membership/Login"
]
for path in paths:
    method = "get"
    status_code = "200"

    response_schema = openapi_spec["paths"][path][method]["responses"][status_code][
        "content"
    ]["application/json"]["schema"]

    url = f"https://abc.com{path}"
    response = requests.get(
        url, params={"Username": "XXX", "Password": "XXXX"}
    )
    response_data_api = response.json()
    
    try:
        validate(instance=response_data_api, schema=response_schema)
        print("Response is valid.")
    except ValidationError as e:
        print("Response is invalid:", e.json_path + e.message)
```

Currently my endpoint is returning an error (intentially) 

```
{'ErrorCode': 1, 'Result': 'ServerSideError'}
```

This response is being passed into `response_data_api` and `response_schema` looks like - 
![image](https://github.com/user-attachments/assets/daa48a81-0825-4248-b635-6dbb8031ef3c)

I can see the schema says it shouled be an object with keys -
ErrorCode = int64 
VehicleTripPeriodDistances = array

Where the current response is 
ErrorCode = int64 
Result = string

But there is no ValidationError being thrown, and I don't understand why, As I can see it fail for other paths of the schema I have tried..

Hope someone can enlighten me on what i am missing...

Thanks

**Comment by Julian** [comment]
Hey there. Minimizing this to remove the web requests and just leaving a simple call to `jsonschema.validate` with a minimal schema and instance would make this lots easier to answer!

**Reply by karl0ss** [reply]
Ok, I will simplify this today and repost, thanks for your response...

**Comment by karl0ss** [comment]
Ok, I have made variables of the response and the schema taken from the openapi spec, and this still passes saying the response is valid 


```
from jsonschema import validate, ValidationError

schema = {'type': 'object', 'properties': {'ErrorCode': {'type': 'integer', 'format': 'int64', 'example': 0}, 'VehicleTripPeriodDistances': {'type': 'array', 'items': {'type': 'object', 'properties': {'VehicleNodeId': {'type': 'integer', 'format': 'int64', 'example': 566}, 'VehicleRegistration': {'type': 'string', 'example': None}, 'TotalPeriodDistance': {'type': 'integer', 'format': 'int64', 'example': 0}}}}}}

response = {'ErrorCode': 2, 'Result': 'SessionExpired'}

try:
    validate(instance=response, schema=schema)
    print("Response is valid.")
except ValidationError as e:
    print("Response is invalid:", e.json_path + e.message)
```

**Reply by Julian** [reply]
Thanks -- and can you say why you expect that to be invalid?

**Reply by karl0ss** [reply]
Now unless i'm missunderstanding what this should be doing, the schema implies..

```
object {
    ErrorCode = integer,
    VehicleTripPeriodDistances = array 
        array [
            object {
                VehicleNodeId = integer,
                 .....
             }
        ]
   }
```

where as the response is 

```
object {
    ErrorCode = integer,
    Result = string
}
```

My understanding would be that this should fail due to the schema implying a few things that aren't correct?

Top level object should have 2 keys, `ErrorCode` and `VehicleTripPeriodDistances` where as we have `ErrorCode` and `Result`?
     a. should this not fail due to VehicleTripPeriodDistances being missing?
     b, should this not fail due to Result being present?

Thanks

Karl

**Reply by Julian** [reply]
No – in JSON Schema properties are not required by default, and additional properties are not disallowed by default -- the former is e.g. the [`required` keyword](https://www.learnjsonschema.com/2020-12/validation/required/), the latter is e.g. the [`additionalProperties` keyword](https://www.learnjsonschema.com/2020-12/applicator/additionalproperties/).

**Reply by karl0ss** [reply]
Ah interesting, seems I need to do some more reading, thank you for your time @Julian

**Reply by Julian** [reply]
Pleasure!
